nighthawk/include/nighthawk/common/sequencer.h

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <functional>
+
+#include "envoy/common/pure.h"
+
+#include "nighthawk/common/statistic.h"
+
+namespace Nighthawk {
+
+using SequencerTarget = std::function<bool(std::function<void()>)>;
+
+/**
+ * Abstract Sequencer interface. The Sequencer will drive calls to the SequencerTarget.
+ * The contract with the target is that it will call the provided callback when it is ready.
+ * The target will return true if it was able to proceed, or false if a retry is warranted at
+ * a later time (because of being out of required resources, for example).
+ */
+class Sequencer {
+public:
+  Sequencer(SequencerTarget& target) : target_(target) {}
+  virtual ~Sequencer() = default;
+
+  /**
+   * Starts the sequencer.
+   */
+  virtual void start() PURE;
+
+  /**
+   * Wait until the sequencer has finished.
+   */
+  virtual void waitForCompletion() PURE;
+
+  /**
+   * @return double an up-to-date completions per second rate.
+   */
+  virtual double completions_per_second() PURE;
```

Comment:
Nit: prefer `completionsPerSecond()` (and same below) for abstract interface methods. Also, some of these could probably be `const`, since they shouldn't need to mutate, e.g. `virtual double completionsPerSecond() const PURE;`